### PR TITLE
add checks when db is opened in read-only mode

### PR DESCRIPTION
**Pull request by NamanJain8**
When db is opened in read-only mode, we should not create/truncate memtables or wal files. This PR addresses that issue.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1565)

### CODE REVIEW

**Review by manishrjain — APPROVED** [review]
:lgtm: Really test thoroughly for any side effects.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1565#-:-MKjO7ABEyz4Y37oumfl:b-kisyir)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @jarifibrahim)

File: db.go [hunk 1]
```diff
@@ -530,6 +535,11 @@ func (db *DB) close() (err error) {
 	db.stopMemoryFlush()
 	db.stopCompactions()
 
+	// Delete the mem table if not already deleted.
+	if db.mt != nil {
```
Comment (jarifibrahim) [inline db.go:539]: why do you need this?

File: db_test.go [hunk 2]
```diff
@@ -2188,3 +2188,24 @@ func TestUpdateMaxCost(t *testing.T) {
...
 }
+
+func TestOpenDB(t *testing.T) {
+	dir, err := ioutil.TempDir("", "badger-test")
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	ops := getTestOptions(dir)
+	ops.ReadOnly = false
+	db, err := Open(ops)
```
Comment (jarifibrahim) [inline db_test.go:2199]: Write some data which goes to sst and vlog file.

File: value_test.go [hunk 3]
```diff
@@ -663,7 +663,7 @@ func TestReadOnlyOpenWithPartialAppendToWAL(t *testing.T) {
 		{Key: k2, Value: v2},
 	})
 	buf = buf[:offset-6]
-	require.NoError(t, ioutil.WriteFile(vlogFilePath(dir, 1), buf, 0777))
```
Comment (jarifibrahim) [inline value_test.go:666]: This test is returning error for memtable even when we have not written any data. Please fix this @NamanJain8

**Review by jarifibrahim — APPROVED** [review]
Test the code by Inserting 1 million records and reopening the db multiple times (without closing the db always). 

Feel free to merge once the CI and your manual tests work fine.

File: db.go (hunk 1 above)
Comment (NamanJain8) [reply db.go]: I used this for deleting the created memtable in ReadOnly mode. This is not needed as we are not creating the memtable in upfront now in read-only mode.

File: db_test.go (hunk 2 above)
Comment (NamanJain8) [reply db_test.go:2201]: Done.

File: value_test.go (hunk 3 above)
Comment (NamanJain8) [reply value_test.go:666]: Fixed this.

File: value.go [hunk 4]
```diff
@@ -894,7 +897,7 @@ func (vlog *valueLog) getFileRLocked(vp valuePointer) (*logFile, error) {
 
 	// Check for valid offset if we are reading from writable log.
 	maxFid := vlog.maxFid
-	if vp.Fid == maxFid {
+	if vp.Fid == maxFid && !vlog.opt.ReadOnly {
```
Comment (NamanJain8) [inline value.go:900]: **Explanation**: In read-only mode, we need not check for the `vlog.writableOffset`.

File: db.go [hunk 5]
```diff
@@ -612,17 +617,20 @@ func (db *DB) getMemTables() ([]*memTable, func()) {
 	db.RLock()
 	defer db.RUnlock()
 
-	tables := make([]*memTable, len(db.imm)+1)
```
Comment (jarifibrahim) [inline db.go:615]: Why do we need this change?

File: db.go [hunk 6]
```diff
@@ -501,7 +506,7 @@ func (db *DB) close() (err error) {
 	// and remove them completely, while the block / memtable writer is still
 	// trying to push stuff into the memtable. This will also resolve the value
 	// offset problem: as we push into memtable, we update value offsets there.
-	if !db.mt.sl.Empty() {
+	if !db.opt.ReadOnly {
```
Comment (jarifibrahim) [inline db.go:509]: We might end up pushing an empty memtable. Why push it?

File: db.go [hunk 7]
```diff
@@ -1351,7 +1359,9 @@ func (db *DB) KeySplits(prefix []byte) []string {
 		for _, mt := range memTables {
 			mtSplits(mt)
 		}
-		mtSplits(db.mt)
+		if !db.opt.ReadOnly {
+			mtSplits(db.mt)
```
Comment (jarifibrahim) [inline db.go:1363]: Add a `if nil return` check in `mtsplit`. It looks much cleaner.

File: db_test.go [hunk 8]
```diff
@@ -2188,3 +2188,81 @@ func TestUpdateMaxCost(t *testing.T) {
...
+func TestOpenDBReadOnly(t *testing.T) {
+	dir, err := ioutil.TempDir("", "badger-test")
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	mp := make(map[string][]byte)
+
+	ops := getTestOptions(dir)
+	ops.ReadOnly = false
+	db, err := Open(ops)
```
Comment (jarifibrahim) [inline db_test.go:2201]: Check err

File: db_test.go [hunk 9]
```diff
@@ -2188,3 +2188,81 @@ func TestUpdateMaxCost(t *testing.T) {
...
+		val := make([]byte, db.opt.ValueThreshold+10)
+		rand.Read(val)
+		for i := 0; i < 10; i++ {
+			key := fmt.Sprintf("KEY-%05d", i)
+			require.NoError(t, txn.Set([]byte(key), val))
+			mp[key] = val
+		}
+		return nil
+	}))
+	require.NoError(t, err)
```
Comment (jarifibrahim) [inline db_test.go:2214]: This should be one line above.

File: value.go (hunk 4 above)
Comment (jarifibrahim) [reply value.go:900]: Why is this the case? the writable offset should be greater than the read offset always.

File: db.go [hunk 10]
```diff
@@ -298,9 +298,11 @@ func Open(opt Options) (*DB, error) {
 		return nil, y.Wrapf(err, "while opening memtables")
 	}
 
-	db.mt, err = db.newMemTable()
-	if err != nil {
-		return nil, y.Wrapf(err, "cannot create memtable")
+	if !db.opt.ReadOnly {
+		db.mt, err = db.newMemTable()
+		if err != nil {
```
Comment (jarifibrahim) [inline db.go:303]: Single line please. `if db.mt, err = db.newMemTable(); err != nil `

**Review by jarifibrahim — COMMENTED** [review]
Got some comments

File: db.go (hunk 6 above)
Comment (NamanJain8) [reply db.go:509]: `handleFlushTask` do nothing on empty memtable. Also, DecrRef() is called hence deleting the empty memtable. So, I am avoiding special handling here.

File: db.go (hunk 5 above)
Comment (NamanJain8) [reply db.go:615]: In readOnly mode, we should not append the mutable memtable in the tables list. The `+1` is for the `db.mt`.
These `tables` get iterated upon and will cause panic on nil tables. `nil` check at various places can be cumbersome as well as untidy. 
So, either we make this change, or create a slice of size `len(db.imm)` in readOnly mode.

File: db.go (hunk 7 above)
Comment (NamanJain8) [reply db.go:1363]: thanks. done.

File: db.go (hunk 10 above)
Comment (NamanJain8) [reply db.go:303]: done.

File: value.go (hunk 4 above)
Comment (NamanJain8) [reply value.go:900]: Yes, it should be. We would not need this change if we update the writable offset in read-only mode as well by iterating over last vlog file.
I was thinking along the lines that this check is for the vlog file that is currently in use(i.e., they may be written to). As we are not checking the validity of offset for the files other than that with the `maxFid`.

File: db.go (hunk 5 above)
Comment (jarifibrahim) [reply db.go:615]: okay

File: value.go [hunk 11]
```diff
@@ -561,7 +561,7 @@ func (vlog *valueLog) open(db *DB) error {
 		return err
 	}
 	// If no files are found, then create a new file.
-	if len(vlog.filesMap) == 0 {
+	if len(vlog.filesMap) == 0 && !vlog.opt.ReadOnly {
```
Comment (jarifibrahim) [inline value.go:564]: This wouldn't work if the db is opened for the first time in read only mode.

**Review by jarifibrahim — APPROVED** [review]

File: db.go (hunk 6 above)
Comment (NamanJain8) [reply db.go]: done

**Pull request merged**